packages/docs/docs/docs/proxies.md

```diff
@@ -0,0 +1,68 @@
+
+## Upgradeability via the proxy pattern
+
+The basic idea behind using the proxy pattern to solve the problem of pugradeability, is to use two contracts instead of one. The first contract is a simple wrapper or "proxy" which users interact with directly and is in charge of forwarding transactions to and from the second contract, which contains the logic of the pair. The key concept to understand is that the logic contract can be replaced while the proxy is never changed. Both contracts are still immutable, in the sense that their code cannot be changed, but the logic contract can simply be swapped by another. The wrapper can thus point to a different logic implementation and in doing so, the software is "upgraded".
+
+## Proxy forwarding
+
+The most immediate problem that proxies need to solve is how the proxy exposes the entire interface of the logic contract without requiring a one to one mapping of the entire logic contract's interface. This would be difficult to maintain, prone to errors, and would make the interface itself not upgradeable. Hence, a dynamic forwarding mechanism is required. The basics of such mechanism can be summarized in the code below:
+
+```
+assembly {
+    let ptr := mload(0x40)
+    calldatacopy(ptr, 0, calldatasize)
+    let result := delegatecall(gas, _impl, ptr, calldatasize, 0, 0)
+    let size := returndatasize
+    returndatacopy(ptr, 0, size)
+
+    switch result
+    case 0 { revert(ptr, size) }
+    default { return(ptr, size) }
+ }
+ ```
+
+This code can be put in the fallback function of a proxy, and will forward any call to any function with any set of parameters to the logic contract without it needing to know anything in particular of the logic contract's interface. In essence, (1) the `calldata` is copied to memory, (2) the call is forwarded to the logic contract, (3) the return data from the call to the logic contract is retrieved, and (4) the returned data is forwarded back to the caller.
+
+A very important thing to note is that the code makes use of `delegatecall` which executes the callee's code in the context of the caller's state. That is, the logic contract controls the proxy's state and the logic contract's state is meaningless. Thus, the proxy doesn't forward transactions to and from the logic contract, but also represents the pair's state. The state is in the proxy and the logic is in the particular implementation that the proxy points to. To put it in yet another way, the "what" is in the proxy contract and the "how" is in the logic contract.
```

Comment:
Link to delegatecall's docs.
doesn't forward -> doesn't only forward

Reply:
I'm not sure I like the last sentence about "what" and "how". What would you think of mentioning instead that the state variables are in the proxy, and the functions are in the logic?

Reply:
I don't like that because it may confuse the user. One may think that variables need to be declared in the proxy in <code> and functions in the implementation.

Removing my sentence and replacing it with nothing for now.